### ISSUE TITLE
Build issue `conda.sh`

### ISSUE THREAD

**Issue by darribas** [conjuring/conjuring#36]
Next step fail:

```docker
Step 7/14 : RUN /conda.sh install -c conda-forge -q -y pip  && /conda.sh path_exec pip install --no-cache-dir -U pip
 ---> Running in d68b2a00d100
/usr/bin/env: ‘bash\r’: No such file or directory
ERROR: Service 'core' failed to build: The command '/bin/sh -c /conda.sh install -c conda-forge -q -y pip  && /conda.sh path_exec pip install --no-cache-dir -U pip' returned a non-zero code: 127
```

**Comment by darribas** [comment]
This error is blocking not only the build but the running. I have no idea why or where it comes from, I didn't encounter it on earlier builds...

**Comment by casperdcl** [comment]
are you using a base image where you've messed with system paths? this error seems to imply `bash` can't be found

**Comment by darribas** [comment]
`base.Dockerfile`:

```Dockerfile
FROM darribas/gds:3.0 as base
USER root
#FROM ubuntu:18.04 as base
```

This has worked before. The `\r` at the end makes me wonder whether it's a Windows mess up on my side while editing one of the files?

**Comment by casperdcl** [comment]
could be but my tests don't seem to indicate line endings should be a problem here.

You could try to save files in linux (LF) rather than win (CRLF) format, but doubt it'll fix this

**Comment by casperdcl** [comment]
does changing back to the normal base (`ubuntu:18.04`) fix things? Wishful thinking :)

**Comment by darribas** [comment]
If I drop into the base image (`darribas/gds:3.0`) and run `src/conda.sh`, I get the same error:

```bash
jovyan@72cca18102c8:~/work$ cat src/conda.sh
#!/usr/bin/env bash
## Usage: conda.sh <cmd> ...
# Where <cmd> is any valid conda command, or `path_exec` (which will prefix
# the base conda bin path before running the remaining command)

if [ ${#} -gt 1 -a "$1" = path_exec ]; then
  PATH="$($0 info --base)/bin:$PATH" "${@:2}"
else
  which conda
  if [ $? -eq 0 ]; then
    conda $@
  else
    /opt/conda/bin/conda $@
  fi
fi
jovyan@72cca18102c8:~/work$
```

**Comment by darribas** [comment]
Also, reading in linux Python doesn't show any `\r`:

```bash
jovyan@72609348fac8:~/work$ ipython
Python 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.7.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: open('src/conda.sh').read()
Out[1]: '#!/usr/bin/env bash\n## Usage: conda.sh <cmd> ...\n# Where <cmd> is any valid conda command, or `path_exec` (which will prefix\n# the base conda bin path before running the remaining command)\n\nif [ ${#} -gt 1 -a "$1" = path_exec ]; then\n  PATH="$($0 info --base)/bin:$PATH" "${@:2}"\nelse\n  which conda\n  if [ $? -eq 0 ]; then\n    conda $@\n  else\n    /opt/conda/bin/conda $@\n  fi\nfi\n'

In [2]:
```

**Comment by casperdcl** [comment]
python modifies line endings. you might need `open('src/conda.sh', mode='rb')` or similar

**Comment by darribas** [comment]
AHAH!!!

```python
In [2]: open("src/conda.sh", 'rb').read()
Out[2]: b'#!/usr/bin/env bash\r\n## Usage: conda.sh <cmd> ...\r\n# Where <cmd> is any valid conda command, or `path_exec` (which will prefix\r\n# the base conda bin path before running the remaining command)\r\n\r\nif [ ${#} -gt 1 -a "$1" = path_exec ]; then\r\n  PATH="$($0 info --base)/bin:$PATH" "${@:2}"\r\nelse\r\n  which conda\r\n  if [ $? -eq 0 ]; then\r\n    conda $@\r\n  else\r\n    /opt/conda/bin/conda $@\r\n  fi\r\nfi\r\n'

In [3]: 
```

How can I remove that?

**Comment by darribas** [comment]
Within Windows that is...

**Comment by darribas** [comment]
We should definitely be supporting Windows-edited files if we are to reach an audience...

**Comment by jreades** [comment]
Looks like we might need to play with two things to ensure that Windows users have an apparently ’seamless’ experience editing files and running docker-compose:

1. Create a .gitattributes files as part of installation in which we set the options below (but to which we should add .py files).
2. We might need to also set a global configuration option (though #1 should override this?) also outlined below, though it seems like this shouldn’t be needed if the initial clone has attributes set.

In .gitattributes:
* text=auto*.sh text eol=lf

Global git setting:

git config --global core.autocrlf input

References:

https://techblog.dorogin.com/case-of-windows-line-ending-in-bash-script-7236f056abe

https://git-scm.com/book/tr/v2/Customizing-Git-Git-Configuration#Formatting-and-Whitespace

https://help.github.com/en/articles/configuring-git-to-handle-line-endings#refreshing-a-repository-after-changing-line-endings

**Comment by casperdcl** [comment]
> You could try to save files in linux (LF) rather than win (CRLF) format, but doubt it'll fix this

my tests show `bash` should be fine with windows line endings.

**Comment by casperdcl** [comment]
@darribas can you confirm your issue is still not solved?

To clairfy, I believe

- #38 does NOT fix this issue
- using a text editor in windows which supports saving with linux-style line endings also does NOT solve this issue

**Comment by szschaler** [comment]
I don’t think this would change the actual line endings if the file is edited on that machine.

Steffen

Dr. rer. nat. Steffen Zschaler AHEA

Senior Lecturer
King's College London
Department of Informatics

Visiting Scientist
The Francis Crick Institute

Email szschaler@acm.org
Phone +44 (020) 7848 1513
WWW   www.steffen-zschaler.de

**Comment by casperdcl** [comment]
no it wouldn't; it only affects commits

**Comment by darribas** [comment]
I "solved" it by moving to WSL in Linux, which is well suboptimal, but got it working for now.

**Comment by casperdcl** [comment]
ok, I'm really skeptical that this solves issue #35 as my tests show `bash` is fine with win as well as mixed line endings. But if it works for you then feel free to close this issue!
